File: src/go/delta/correcting.go
package delta

import "fmt"
import "os"

// diffCorrecting implements the Correcting 1.5-Pass algorithm (Section 7, Figure 8)
// with fingerprint-based checkpointing (Section 8).
//
// |C| = cap (hash table capacity, auto-sized from input).
// |F| = next_prime(2 * num_R_seeds) (footprint universe, Section 8.1).
// m  = ceil(|F| / |C|) (checkpoint spacing, p. 348).
// k  = checkpoint class (Eq. 3, p. 348).
func diffCorrecting(r, v []byte, opts DiffOptions) []Command {
	if len(v) == 0 {
		return nil
	}

	p := opts.P
	q := opts.Q
	bufCap := opts.BufCap
	verbose := opts.Verbose
	useSplay := opts.UseSplay

	// ── Checkpointing parameters (Section 8.1, pp. 347-348) ──
	numSeeds := 0
	if len(r) >= p {
		numSeeds = len(r) - p + 1
	}
	maxTable := opts.MaxTable
	if maxTable <= 0 {
		maxTable = MaxTableSize
	}

	var cap_ int
	if numSeeds > 0 {
		twoSeedsOverP := 2 * numSeeds / p
		cap_ = int(NextPrime(int64(min64(int64(maxTable), max64(int64(q), int64(twoSeedsOverP))))))
	} else {
		cap_ = int(NextPrime(int64(min64(int64(q), int64(maxTable)))))
	}

	fSize := int64(1)
	if numSeeds > 0 {
		fSize = NextPrime(int64(2 * numSeeds))
	}
	m := int64(1)
	if fSize > int64(cap_) {
		m = (fSize + int64(cap_) - 1) / int64(cap_)
	}

	k := int64(0)
	if len(v) >= p {
		mid := len(v) / 2
		if mid > len(v)-p {
			mid = len(v) - p
		}
		k = Fingerprint(v, mid, p) % fSize % m
	}

	if verbose {
		structName := "hash table"
		if useSplay {
			structName = "splay tree"
		}
		var expected int64
		if m > 0 {
			expected = int64(numSeeds) / m
		}
		var occEst int64
		if cap_ > 0 {
			occEst = expected * 100 / int64(cap_)
		}
		fmt.Fprintf(os.Stderr,
			"correcting: %s, |C|=%d |F|=%d m=%d k=%d\n"+
				"  checkpoint gap=%d bytes, expected fill ~%d (~%d%% table occupancy)\n",
			structName, cap_, fSize, m, k, m, expected, occEst)
	}

	// Step (1): build lookup structure for R (first-found / insert-if-absent policy).
	var htFp []int64
	var htOff []int
	var splayR *SplayTree[[2]int64] // value = [full_fp, offset]

	if useSplay {
		splayR = &SplayTree[[2]int64]{}
	} else {
		htFp = make([]int64, cap_)
		for i := range htFp { htFp[i] = -1 }
		htOff = make([]int, cap_)
	}

	if numSeeds > 0 {
		rhR := NewRollingHash(r, 0, p)
		for a := 0; a < numSeeds; a++ {
			var fp int64
			if a == 0 {
				fp = rhR.Value()
			} else {
				rhR.Roll(int(r[a-1]&0xFF), int(r[a+p-1]&0xFF))
				fp = rhR.Value()
			}
			f := fp % fSize
			if f%m != k {
				continue // not a checkpoint seed
			}
			if useSplay {
				_, ok := splayR.Find(fp)
				if !ok {
					splayR.Insert(fp, [2]int64{fp, int64(a)})
				}
			} else {
				i := int(f / m)
				i0 := i
				for {
					if htFp[i] == -1 { break }          // empty — store here
					if htFp[i] == fp { i = -1; break }  // dup fp — skip
					i++
					if i == cap_ { i = 0 }
					if i == i0 { i = -1; break }         // table full
				}
				if i >= 0 {
					htFp[i] = fp
					htOff[i] = a
				}
			}
		}
	}

	if verbose && numSeeds > 0 {
		// Build phase stats.
		storedCount := 0
		if useSplay {
			storedCount = splayR.Len()
		} else {
			for _, fp := range htFp {
				if fp != -1 {
					storedCount++
				}
			}
		}
		// We'll compute exact counts by re-scanning; just emit table occupancy.
		tableOcc := int64(storedCount)
		fmt.Fprintf(os.Stderr, "  build: table occupancy %d/%d (%.1f%%)\n",
			tableOcc, cap_, float64(tableOcc)*100/float64(cap_))
	}

	// Lookback buffer (Section 5.2).
	//
	// The correcting algorithm may discover that a newly found match overlaps
	// commands already emitted. The buffer holds the most recent bufCap tentative
	// commands so they can be trimmed or cancelled (tail correction) when a better
	// match is found. Commands are flushed to the output list as they age out.
	type bufEntry struct {
		vStart int     // First V byte covered by this entry.
		vEnd   int     // One past the last V byte covered.
		cmd    Command // The tentative command (CopyCmd or AddCmd).
		dummy  bool    // Reserved; always false in the current implementation.
	}
	buf := make([]bufEntry, 0, bufCap)

	// Step (2): initialize scan pointers.
	vC, vS := 0, 0
	var rhV *RollingHash
	rhVPos := 0
	vSeeds := 0
	if len(v) >= p {
		vSeeds = len(v) - p + 1
		rhV = NewRollingHash(v, 0, p)
	}
	_ = vSeeds

	var commands []Command

	emitOldest := func() {
		if len(buf) >= bufCap {
			oldest := buf[0]
			buf = buf[1:]
			if !oldest.dummy {
				commands = append(commands, oldest.cmd)
			}
		}
	}

	for vC+p <= len(v) { // Step (3)
		// Step (4): fingerprint at vC, apply checkpoint test.
		if rhV == nil {
			break
		}
		var fpV int64
		if vC == rhVPos {
			fpV = rhV.Value()
		} else if vC == rhVPos+1 {
			rhV.Roll(int(v[vC-1]&0xFF), int(v[vC+p-1]&0xFF))
			rhVPos = vC
			fpV = rhV.Value()
		} else {
			rhV = NewRollingHash(v, vC, p)
			rhVPos = vC
			fpV = rhV.Value()
		}

		fV := fpV % fSize
		if fV%m != k {
			vC++
			continue
		}

		// Checkpoint passed — look up R.
		var storedFp int64
		var rOffset int
		if useSplay {
			entry, ok := splayR.Find(fpV)
			if !ok {
				vC++
				continue
			}
			storedFp = entry[0]
			rOffset = int(entry[1])
		} else {
			i := int(fV / m)
			i0 := i
			found := -1
			for {
				if htFp[i] == -1 { break }            // empty — chain ends
				if htFp[i] == fpV { found = i; break }
				i++
				if i == cap_ { i = 0 }
				if i == i0 { break }                   // full table — not found
			}
			if found < 0 {
				vC++
				continue
			}
			storedFp = htFp[found]
			rOffset = htOff[found]
		}

		if storedFp != fpV {
			vC++
			continue
		}
		if !regionEquals(r, rOffset, v, vC, p) {
			vC++
			continue
		}

		// Step (5): extend match forwards and backwards.
		fwd := p
		for vC+fwd < len(v) && rOffset+fwd < len(r) && v[vC+fwd] == r[rOffset+fwd] {
			fwd++
		}
		bwd := 0
		for vC >= bwd+1 && rOffset >= bwd+1 && v[vC-bwd-1] == r[rOffset-bwd-1] {
			bwd++
		}

		vM := vC - bwd
		rM := rOffset - bwd
		ml := bwd + fwd
		matchEnd := vM + ml

		if ml < p {
			vC++
			continue
		}

		// Step (6): encode with correction.
		if vS <= vM {
			// (6a) match in unencoded suffix.
			if vS < vM {
				emitOldest()
				data := make([]byte, vM-vS)
				copy(data, v[vS:vM])
				buf = append(buf, bufEntry{vStart: vS, vEnd: vM, cmd: AddCmd{Data: data}})
			}
			emitOldest()
			buf = append(buf, bufEntry{vStart: vM, vEnd: matchEnd, cmd: CopyCmd{Offset: rM, Length: ml}})
			vS = matchEnd
		} else {
			// (6b) tail correction (Section 5.1, p. 339).
			effectiveStart := vS

			for len(buf) > 0 {
				tail := &buf[len(buf)-1]
				if tail.dummy {
					buf = buf[:len(buf)-1]
					continue
				}
				if tail.vStart >= vM && tail.vEnd <= matchEnd {
					if tail.vStart < effectiveStart {
						effectiveStart = tail.vStart
					}
					buf = buf[:len(buf)-1]
					continue
				}
				if tail.vEnd > vM && tail.vStart < vM {
					if _, isAdd := tail.cmd.(AddCmd); isAdd {
						keep := vM - tail.vStart
						if keep > 0 {
							data := make([]byte, keep)
							copy(data, v[tail.vStart:vM])
							tail.cmd = AddCmd{Data: data}
							tail.vEnd = vM
						} else {
							buf = buf[:len(buf)-1]
						}
						if vM < effectiveStart {
							effectiveStart = vM
						}
					}
					break
				}
				break
			}

			adj := effectiveStart - vM
			newLen := matchEnd - effectiveStart
			if newLen > 0 {
				emitOldest()
				buf = append(buf, bufEntry{vStart: effectiveStart, vEnd: matchEnd,
					cmd: CopyCmd{Offset: rM + adj, Length: newLen}})
			}
			vS = matchEnd
		}

		// Step (7): advance past matched region.
		vC = matchEnd
	}

	// Step (8): flush buffer and trailing add.
	for _, entry := range buf {
		if !entry.dummy {
			commands = append(commands, entry.cmd)
		}
	}
	if vS < len(v) {
		data := make([]byte, len(v)-vS)
		copy(data, v[vS:])
		commands = append(commands, AddCmd{Data: data})
	}

	if verbose {
		printStats(commands)
	}
	return commands
}

func min64(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

File: src/kotlin/delta/Correcting.kt
package delta

/**
 * Correcting 1.5-Pass algorithm (Section 7, Figure 8) with
 * fingerprint-based checkpointing (Section 8).
 *
 * |C| = q (hash table capacity, auto-sized from input).
 * |F| = next_prime(2 * num_R_seeds) (footprint universe, Section 8.1).
 * m  = ceil(|F| / |C|) (checkpoint spacing, p. 348).
 * k  = checkpoint class (Eq. 3, p. 348).
 */

/**
 * One entry in the correction lookback buffer (Section 5.2).
 *
 * The correcting algorithm may discover that a newly found match overlaps
 * commands already emitted.  The buffer holds the most recent bufCap tentative
 * commands so they can be trimmed or cancelled (tail correction) when a better
 * match is found.  Commands are flushed to the output list as they age out.
 *
 * @param vStart First V byte covered by this entry.
 * @param vEnd   One past the last V byte covered.
 * @param cmd    The tentative command (Add or Copy).
 * @param dummy  Reserved; always false in the current implementation.
 */
private class BufEntry(val vStart: Int, var vEnd: Int, var cmd: Command, val dummy: Boolean)

fun diffCorrecting(r: ByteArray, v: ByteArray, opts: DiffOptions): List<Command> {
    val commands = mutableListOf<Command>()
    if (v.isEmpty()) return commands

    val p        = opts.p
    val bufCap   = opts.bufCap
    val verbose  = opts.verbose
    val useSplay = opts.useSplay

    // ── Checkpointing parameters (Section 8.1, pp. 347-348) ─────────────────
    val numSeeds = if (r.size >= p) r.size - p + 1 else 0
    val maxTable = if (opts.maxTable > 0) opts.maxTable else MAX_TABLE_SIZE
    val cap = if (numSeeds > 0)
        nextPrime(minOf(maxTable.toLong(), maxOf(opts.q.toLong(), 2L * numSeeds / p))).toInt()
    else
        nextPrime(minOf(opts.q.toLong(), maxTable.toLong())).toInt()
    val fSize: Long = if (numSeeds > 0) nextPrime(2L * numSeeds) else 1L
    val m: Long     = if (fSize <= cap) 1L else (fSize + cap - 1) / cap
    val k: Long     = if (v.size >= p)
        fingerprint(v, minOf(v.size / 2, v.size - p), p) % fSize % m
    else 0L

    if (verbose) {
        val expected = if (m > 0) numSeeds / m else 0L
        val occEst   = if (cap > 0) expected * 100 / cap else 0L
        System.err.printf(
            "correcting: %s, |C|=%d |F|=%d m=%d k=%d%n" +
            "  checkpoint gap=%d bytes, expected fill ~%d (~%d%% table occupancy)%n",
            if (useSplay) "splay tree" else "hash table", cap, fSize, m, k,
            m, expected, occEst)
    }

    // Step (1): Build lookup structure for R (first-found policy)
    val htFp:   LongArray?
    val htOff:  IntArray?
    val splayR: SplayTree<LongArray>?  // value = [full_fp, offset]

    if (useSplay) {
        splayR = SplayTree()
        htFp = null; htOff = null
    } else {
        splayR = null
        htFp   = LongArray(cap) { -1L }
        htOff  = IntArray(cap)
    }

    var rhR: RollingHash? = if (numSeeds > 0) RollingHash(r, 0, p) else null
    for (a in 0 until numSeeds) {
        val fp: Long = if (a == 0) {
            rhR!!.value
        } else {
            rhR!!.roll(r[a - 1].toInt() and 0xFF, r[a + p - 1].toInt() and 0xFF)
            rhR.value
        }
        val f = fp % fSize
        if (f % m != k) continue  // not a checkpoint seed

        if (useSplay) {
            splayR!!.insertOrGet(fp, longArrayOf(fp, a.toLong()))
        } else {
            var i = (f / m).toInt()
            val i0 = i
            while (true) {
                if (htFp!![i] == -1L) { break }               // empty — store here
                if (htFp!![i] == fp) { i = -1; break }        // dup fp — skip
                if (++i == cap) { i = 0 }
                if (i == i0) { i = -1; break }                 // table full
            }
            if (i >= 0) { htFp!![i] = fp; htOff!![i] = a }
        }
    }

    // Encoding lookback buffer (Section 5.2)
    val buf = ArrayDeque<BufEntry>()

    // Step (2): initialize scan pointers
    var vC    = 0
    var vS    = 0
    val vSeeds = if (v.size >= p) v.size - p + 1 else 0
    var rhV   = if (vSeeds > 0) RollingHash(v, 0, p) else null
    var rhVPos = 0

    while (vC + p <= v.size) {  // Step (3): check for end of V
        // Step (4): fingerprint at vC, apply checkpoint test.
        val rh = rhV ?: break
        val fpV: Long = when {
            vC == rhVPos     -> rh.value
            vC == rhVPos + 1 -> {
                rh.roll(v[vC - 1].toInt() and 0xFF, v[vC + p - 1].toInt() and 0xFF)
                rhVPos = vC; rh.value
            }
            else -> {
                val newRh = RollingHash(v, vC, p); rhV = newRh; rhVPos = vC; newRh.value
            }
        }

        val fV = fpV % fSize
        if (fV % m != k) { vC++; continue }

        // Checkpoint passed — look up R.
        val storedFp: Long
        val rOffset: Int
        if (useSplay) {
            val entry = splayR!!.find(fpV)
            if (entry == null) { vC++; continue }
            storedFp = entry[0]; rOffset = entry[1].toInt()
        } else {
            var i = (fV / m).toInt()
            val i0 = i
            var found = -1
            while (true) {
                if (htFp!![i] == -1L) { break }               // empty — chain ends
                if (htFp!![i] == fpV) { found = i; break }
                if (++i == cap) { i = 0 }
                if (i == i0) { break }                         // full table — not found
            }
            if (found < 0) { vC++; continue }
            storedFp = htFp!![found]; rOffset = htOff!![found]
        }

        if (storedFp != fpV) { vC++; continue }
        if (!r.regionEquals(rOffset, v, vC, p)) { vC++; continue }

        // Step (5): extend match forwards and backwards
        var fwd = p
        while (vC + fwd < v.size && rOffset + fwd < r.size && v[vC + fwd] == r[rOffset + fwd]) fwd++
        var bwd = 0
        while (vC >= bwd + 1 && rOffset >= bwd + 1 && v[vC - bwd - 1] == r[rOffset - bwd - 1]) bwd++

        val vM = vC - bwd
        val rM = rOffset - bwd
        val ml = bwd + fwd
        val matchEnd = vM + ml

        if (ml < p) { vC++; continue }

        // Step (6): encode with correction
        if (vS <= vM) {
            // (6a) match in unencoded suffix
            if (vS < vM) {
                if (buf.size >= bufCap) {
                    val oldest = buf.removeFirst()
                    if (!oldest.dummy) commands.add(oldest.cmd)
                }
                buf.addLast(BufEntry(vS, vM, Command.Add(v.copyOfRange(vS, vM)), false))
            }
            if (buf.size >= bufCap) {
                val oldest = buf.removeFirst()
                if (!oldest.dummy) commands.add(oldest.cmd)
            }
            buf.addLast(BufEntry(vM, matchEnd, Command.Copy(rM, ml), false))
            vS = matchEnd
        } else {
            // (6b) tail correction (Section 5.1, p. 339)
            var effectiveStart = vS

            while (buf.isNotEmpty()) {
                val tail = buf.last()
                if (tail.dummy) { buf.removeLast(); continue }

                if (tail.vStart >= vM && tail.vEnd <= matchEnd) {
                    effectiveStart = minOf(effectiveStart, tail.vStart)
                    buf.removeLast(); continue
                }

                if (tail.vEnd > vM && tail.vStart < vM) {
                    if (tail.cmd is Command.Add) {
                        val keep = vM - tail.vStart
                        if (keep > 0) {
                            tail.cmd = Command.Add(v.copyOfRange(tail.vStart, vM))
                            tail.vEnd = vM
                        } else {
                            buf.removeLast()
                        }
                        effectiveStart = minOf(effectiveStart, vM)
                    }
                    break
                }
                break
            }

            val adj    = effectiveStart - vM
            val newLen = matchEnd - effectiveStart
            if (newLen > 0) {
                if (buf.size >= bufCap) {
                    val oldest = buf.removeFirst()
                    if (!oldest.dummy) commands.add(oldest.cmd)
                }
                buf.addLast(BufEntry(effectiveStart, matchEnd, Command.Copy(rM + adj, newLen), false))
            }
            vS = matchEnd
        }

        // Step (7): advance past matched region
        vC = matchEnd
    }

    // Step (8): flush buffer and trailing add
    for (entry in buf) {
        if (!entry.dummy) commands.add(entry.cmd)
    }
    if (vS < v.size) commands.add(Command.Add(v.copyOfRange(vS, v.size)))

    if (verbose) printStats(commands)
    return commands
}
